prov/lnx/src/lnx_ep.c

```diff
@@ -0,0 +1,1188 @@
+}
+
+int lnx_getname(fid_t fid, void *addr, size_t *addrlen)
+{
+	struct local_prov *entry;
+	size_t size = sizeof(struct lnx_addresses);
+	/* initial location to put the address */
+	char ep_addr[FI_NAME_MAX];
+	char *tmp = NULL;
+	struct lnx_addresses *la;
+	struct lnx_address_prov *lap;
+	char hostname[FI_NAME_MAX];
+	size_t prov_addrlen;
+	size_t addrlen_list[LNX_MAX_LOCAL_EPS];
+	int rc, j = 0;
+	struct lnx_ep *lnx_ep;
+	struct lnx_fabric *fabric;
+	struct local_prov_ep *ep;
+
+	lnx_ep = container_of(fid, struct lnx_ep, le_ep.ep_fid.fid);
+	fabric = lnx_ep->le_domain->ld_fabric;
+
+	/* check the hostname and compare it to mine
+	 * TODO: Is this good enough? or do we need a better way of
+	 * determining if the address is local?
+	 */
+	rc = gethostname(hostname, FI_NAME_MAX);
+	if (rc == -1) {
+		FI_WARN(&lnx_prov, FI_LOG_CORE, "failed to get hostname\n");
+		return -FI_EPERM;
+	}
+
+	addrlen_list[0] = 0;
+
+	/* calculate the size of the address */
+	dlist_foreach_container(&fabric->local_prov_table, struct local_prov,
+				entry, lpv_entry) {
+		size += sizeof(struct lnx_address_prov);
+		prov_addrlen = 0;
+
+		dlist_foreach_container(&entry->lpv_prov_eps,
+			struct local_prov_ep, ep, entry) {
+			rc = fi_getname(&ep->lpe_ep->fid, (void*)ep_addr, &prov_addrlen);
+			if (rc == -FI_ETOOSMALL) {
+				size += prov_addrlen * entry->lpv_ep_count;
+				addrlen_list[j] = prov_addrlen;
+				j++;
+				break;
+			} else {
+				/* this shouldn't have happened. */
+				return -FI_EINVAL;
+			}
+		}
+	}
+
+	if (!addr || *addrlen < size) {
+		*addrlen = size;
+		return -FI_ETOOSMALL;
+	}
+
+	la = addr;
+
+	lap = (struct lnx_address_prov *)((char*)la + sizeof(*la));
+
+	j = 0;
+	dlist_foreach_container(&fabric->local_prov_table, struct local_prov,
+							entry, lpv_entry) {
+		memcpy(lap->lap_prov, entry->lpv_prov_name, FI_NAME_MAX - 1);
+		lap->lap_addr_count = entry->lpv_ep_count;
+		lap->lap_addr_size = addrlen_list[j];
+
+		dlist_foreach_container(&entry->lpv_prov_eps,
+			struct local_prov_ep, ep, entry) {
+			tmp = (char*)lap + sizeof(*lap);
+
+			rc = fi_getname(&ep->lpe_ep->fid, (void*)tmp, &addrlen_list[j]);
+			if (rc)
+				return rc;
+
+			if (lap->lap_addr_size != addrlen_list[j])
+				return -FI_EINVAL;
+
+			tmp += addrlen_list[j];
+		}
+
+		lap = (struct lnx_address_prov *)tmp;
+		j++;
+	}
+
+	la->la_prov_count = j;
+	memcpy(la->la_hostname, hostname, FI_NAME_MAX - 1);
+
+	return 0;
+}
+
+static ssize_t lnx_ep_cancel(fid_t fid, void *context)
+{
+	int rc = 0;
+	struct lnx_ep *lep;
+	struct lnx_ctx *ctx;
+	struct local_prov_ep *ep;
+	struct local_prov *entry;
+	struct lnx_fabric *fabric;
+
+	switch (fid->fclass) {
+	case FI_CLASS_EP:
+		lep = container_of(fid, struct lnx_ep, le_ep.ep_fid.fid);
+		break;
+	case FI_CLASS_RX_CTX:
+		ctx = container_of(fid, struct lnx_ctx, ctx_ep.fid);
+		lep = ctx->ctx_parent;
+		break;
+	case FI_CLASS_TX_CTX:
+		/* can't cancel a transmit */
+		return -FI_ENOENT;
+	default:
+		return -FI_EINVAL;
```

Comment:
Is canceling entries on the lnx_peer_srq supported?

Reply:
It forwards the cancel to the core providers and it's up to the core provider to support it or not. The current code doesn't fail when the core providers don't support them

Reply:
I'll look over the rx path again and get back to this.